File: Python/make_two_queue_sum_eqaul.py
from collections import deque

def solution(que1, que2):
    result = 0
    sum1 = 0
    sum2 = 0
    
    que1 = deque(que1)
    que2 = deque(que2)
    
    # 모든 원소의 합 계산
    sum1 = sum(que1)
    sum2 = sum(que2)
    total_sum = sum1 + sum2
        
    # 어떤 방법으로도 각 큐의 원소 합을 같게 만들 수 없는 경우, -1을 return
    if (total_sum % 2 != 0):
        return -1
    total_sum /= 2
    
    if sum1 == sum2:
        return 0
    
    while que1 and que2:
        if sum1 == total_sum:
            return result
        
        elif sum1 > total_sum:
            top = que1.popleft()
            sum1 -= top
            result += 1
        else:
            top = que2.popleft()
            que1.append(top)
            sum1 += top
            result += 1
            
    return -1

File: Swift/make_two_queue_sum_eqaul.swift
import Foundation

class Queue<T> {
    var enQueue: [T]
    var deQueue: [T] = []
    
    var count: Int {
        return enQueue.count + deQueue.count
    }
    
    var isEmpty: Bool {
        return enQueue.isEmpty && deQueue.isEmpty
    }
    
    init(_ queue: [T]) {
        self.enQueue = queue
    }
    
    func push(_ element: T) {
        enQueue.append(element)
    }
    
    func pop() -> T {
        if deQueue.isEmpty {
            deQueue = enQueue.reversed()
            enQueue.removeAll()
        }
        return deQueue.popLast()!
    }
    
}

func solution(_ queue1:[Int], _ queue2:[Int]) -> Int {
    var answer = 0
    var que1 = Queue<Int>(queue1)
    var que2 = Queue<Int>(queue2)
    var sum1 : CLong = queue1.reduce(0, +)
    var sum2 : CLong = queue2.reduce(0, +)
    
    if (sum1 + sum2) % 2 != 0 {
        return -1
    }
    let sum = (sum1 + sum2) / 2
    
    var count = 0
    while (!que1.isEmpty && !que2.isEmpty && count <= 600000) {
        if sum1 < sum {
            let temp = que2.pop()
            que1.push(temp)
            sum1 += temp
            answer += 1
        } else if sum1 > sum {
            let temp = que1.pop()
            que2.push(temp)
            sum1 -= temp
            answer += 1
        }
        if sum1 == sum {
            return answer
        }
        count += 1
    }
    
    if sum1 == sum {
        return answer
    } else {
        return -1
    }
    
    
}
